File: Linked List/linkedList.py


class Node:
    def __init__(self, value=None):
        self.value = value
        self.next = None

    def return_none(self):
        return self.value


class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, value):
        if self.is_empty():
            self.head = value
        else:
            curr = self.head
            while curr.next is not None:
                curr = curr.next

            curr.next = value

    def return_head(self):
        return self.head

    def is_empty(self):
        return not self.head

    def prepend(self, value):
        if self.is_empty():
            self.head = value
        else:
            value.next = self.head
            self.head = value

    def last_index(self):
        current = self.head
        while current.next is not None:
            current = current.next

        return current.value

    def at(self, index):
        i = 0
        curr = self.head
        while i != index:
            if curr.next is None:
                return "Not in linked list"
            curr = curr.next
            i += 1

        return curr.value

    def contains(self, item):
        curr = self.head

        while curr.next is not None:
            if curr.value == item:
                return True
            
            curr = curr.next

        return False

    def find(self, value):
        i = 0
        curr = self.head
        while curr.next is not None:
            if curr.value == value:
                return f"Found at index {i}"

            curr = curr.next
            i += 1

        return None

    def string(self):
        string = ""
        current = self.head
        while current is not None:
            string += f"{current.value}->"
            current = current.next
        string += "None"

        print(string)
        return


linkedList = LinkedList()
new_node = Node(1)

linkedList.append(new_node)
linkedList.prepend(Node(3))
linkedList.append(Node(4))
linkedList.string()
print(linkedList.contains(7))
print(linkedList.last_index())
print(linkedList.find(3))

File: Linked List/linkedList.js

class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}
//
class linkedList {
    constructor() {
        this.head = null;
    }

    append(value) {

        const newNode = new Node(value)

        if (!this.head) {
            this.head = newNode;
        } else {
            let curr = this.head
            while (curr.next != null) {
                curr = curr.next;
            }
        
            curr.next = newNode;
        }
    }

    toString() {
        let string = "[";
        let curr = this.head;
        
        while (curr != null) {
            string += `${curr.value},`
            curr = curr.next;
        }
        string += "]"
        return string

    }

    size() {
        let i = 0;
        let curr = this.head;
        
        while (curr != null) {
            i++;
            curr = curr.next;
        }

        return i;
    }

    head() {
        return this.head;
    }

    tail() {
        let curr = this.head;

        while (curr.next != null) {
            curr = curr.next;
        }

        return curr.value;
    }

    at(index) {
        let i = 0;
        let curr = this.head;

        while (i != index) {
            if (curr.next === null) {
                return "not in linked list";
            }
            curr = curr.next;
            i++;
        }

        return curr.value

    }

    pop() {
        let curr = this.head

        while (curr.next.next != null) {
            curr = curr.next;
        }

        curr.next = null;

        return "Popped";
    }

    contains(value) {
        let curr = this.head;

        if (curr.value === value) {
            return true
        }

        while (curr.next != null) {
            curr = curr.next;
            if (curr.value === value) {
                return true;
            }
        }
        return false;
    }

    find(index) {
        let curr = this.head;
        let i = 0;

        while (curr.value != null) {
            if (curr.value === index) {
                return i;
            }
            curr = curr.next;
            i++;
        }

        return null
    
    }

}


let link = new linkedList()
link.append(2);
link.append(3);
link.append(4);
link.append(5);
link.append(6);
console.log(link.size());
console.log(link.tail());
console.log(link.at(0))
link.pop();
console.log(link.contains(2));
console.log(link.find(5));
console.log(link.toString());